utils/ffmpegvideodecode/ffmpeg_video_dec.cpp

```diff
@@ -0,0 +1,866 @@
+        output_surface_info_.output_surface_size_in_bytes = GetFrameSize();
+        output_surface_info_.mem_type = OUT_SURFACE_MEM_HOST_COPIED;
+    }
+
+    // If the coded_width or coded_height hasn't changed but display resolution has changed, then need to update width and height for
+    // correct output with cropping. There is no need to reconfigure the decoder.
+    if (!is_decode_res_changed && is_display_rect_changed) {
+        return 1;
+    }
+
+    input_video_info_str_.str("");
+    input_video_info_str_.clear();
+    input_video_info_str_ << "Input Video Resolution Changed:" << std::endl
+        << "\tCoded size   : [" << p_video_format->coded_width << ", " << p_video_format->coded_height << "]" << std::endl
+        << "\tDisplay area : [" << p_video_format->display_area.left << ", " << p_video_format->display_area.top << ", "
+            << p_video_format->display_area.right << ", " << p_video_format->display_area.bottom << "]" << std::endl;
+    input_video_info_str_ << std::endl;
+    is_decoder_reconfigured_ = true;
+    return 1;
+}
+
+/**
+ * @brief 
+ * 
+ * @param pPicParams 
+ * @return int 1: success 0: fail
+ */
+int FFMpegVideoDecoder::HandlePictureDecode(RocdecPicParams *pPicParams) {
+    AVPacket *av_pkt = av_packets_[av_pkt_cnt_];
+    std::pair<uint8_t *, int> *packet_data = &av_packet_data_[av_pkt_cnt_];
+    if (last_packet_.payload_size > packet_data->second) {
+        void *new_pkt_data = av_realloc(av_pkt->data, (last_packet_.payload_size + MAX_AV_PACKET_DATA_SIZE));  // add more to avoid frequence reallocation
+        if (!new_pkt_data) {
+            std::cerr << "ERROR: couldn't allocate packet data" << std::endl;
+        }
+        packet_data->first   = static_cast<uint8_t *> (new_pkt_data);
+        packet_data->second  = (last_packet_.payload_size + MAX_AV_PACKET_DATA_SIZE);
+        av_pkt->data = packet_data->first;
+    }
+    memcpy(av_pkt->data, last_packet_.payload, last_packet_.payload_size);
+    av_pkt->size = last_packet_.payload_size;
+    av_pkt->flags = 0;
+    av_pkt->pts = last_packet_.pts;
+
+#if NO_DECODE_THREAD
+    // for testing Decoding without threading
+    DecodeAvFrame(av_pkt, dec_frames_[av_frame_cnt_]);
+#else
+    //push packet into packet q for decoding
+    PushPacket(av_pkt);
+#endif
+    av_pkt_cnt_ = (av_pkt_cnt_ + 1) % av_packets_.size();
+    if (!av_pkt->data || !av_pkt->size) {
+        end_of_stream_ = true;
+    }
+
+    return 1;
+}
+
+/**
+ * @brief function to handle display picture
+ * 
+ * @param pDispInfo 
+ * @return int 0:fail 1: success
+ */
+int FFMpegVideoDecoder::HandlePictureDisplay(RocdecParserDispInfo *pDispInfo) {
+    // check if we reached eos here. This is a hack since HandlePictureDecode won't be called at the end_of_sequece. 
+    // so we need to flush FFMpeg decoder when we have received the lastpacket with 0 bytes
+    if (!last_packet_.payload_size && !end_of_stream_) {
+        AVPacket pkt = { 0 };
+    #if NO_DECODE_THREAD
```

Comment:
I recommend using an environment variable to control whether decoding is performed with or without a thread. This change would enable us to test both options without the need to recompile the program. For instance, we could create an environment variable called `ROCDEC_FFMPEG_DISABLE_DECODE_THREAD`. By setting it with the command `export ROCDEC_FFMPEG_DISABLE_DECODE_THREAD=1`, we can disable threading for FFMPEG. The following code can then be used to check whether threading is disabled.

```
const char *is_threading_disabled = std::getenv("ROCDEC_FFMPEG_DISABLE_DECODE_THREAD");

if (std::string(is_threading_disabled) != "1") 
   DecodeAvFrame(&pkt, dec_frames_[av_frame_cnt_]);
else
 PushPacket(&pkt);
```